tools/ci.sh

```diff
@@ -2,41 +2,153 @@
 ###
 ### Continous Integration Shell Script
 ###
-### This is designed to be the equivalent of what's in .travis.yml, but aimed at executing
+### Designed to be re-entrant on a local dev machine as well, not just on a newly pulled up VM.
 ###
 
-set -euxo pipefail
-
-docker --version
-docker-compose --version
-node --version
-npm --version
-java -version | true
-npm install
-npm run test
-
-cd examples/simple-asset-transfer
-npm install
-rm -rf ./node_modules/websocket/.git/
-rm -rf fabric/api/fabric-client-kv-org*
-cd ./fabric/api/
-rm -rf ./node_modules/websocket/.git/
-npm install
-cd ../../
-npm run fabric:down
-npm run fabric
-cd ./quorum/api/
-rm -rf ./node_modules/websocket/.git/
-npm install
-cd ../../
-npm run quorum:down
-npm run quorum
-npm run fed:build
-npm run fed:quorum
-docker images
-while docker ps | grep "starting\|unhealthy"; do sleep 15; echo; date; done
-docker ps -a
-npm run fabric:down
-npm run quorum:down
-cd ../..
-#  - npm run test:bc - TODO: fix it, broken now
+startedAt=`date +%s`
+baseDir=`pwd`
+
+function mainTask()
+{
+  set -euxo pipefail
+
+  if ! [ -x "$(command -v lscpu)" ]; then
+    echo 'lscpu is not installed, skipping...'
+  else
+    lscpu
+  fi
+
+  if ! [ -x "$(command -v lsmem)" ]; then
+    echo 'lsmem is not installed, skipping...'
+  else
+    lsmem
+  fi
+
+  docker --version
+  docker-compose --version
+  node --version
+  npm --version
+  java -version || true
+
+  ### COMMON
+
+  rm -rf ./examples/simple-asset-transfer/fabric/api/node_modules
+  rm -rf ./examples/simple-asset-transfer/quorum/api/node_modules
+  rm -rf ./examples/simple-asset-transfer/node_modules
+  rm -rf ./node_modules
+
+  npm install
+  npm run test
+
+  cd examples/simple-asset-transfer
+  npm run fed:quorum:down
+  npm run fabric:down
+  npm run quorum:down
+  npm run quorum:api:down
+
+  npm install
+  rm -rf ./node_modules/websocket/.git/
+
+  ### FABRIC
+  cd ./fabric/api/
+  rm -rf ./fabric-client-kv-org*
+  rm -rf ./node_modules/websocket/.git/
+  npm install
+  cd ../../
+  npm run fabric
+
+  ### QUORUM
+
+  # Create docker network if not already in place
+  docker network inspect quorum_quorum-examples-net > /dev/null || docker network create --gateway=172.16.239.1 --subnet=172.16.239.0/24 quorum_quorum-examples-net
+  cd ./quorum/api/
+  rm -rf ./node_modules/websocket/.git/
+  npm install
+  cd ../../
+  npm run quorum
+  npm run quorum:api:build
+  npm run quorum:api
+  npm run fed:build
+  npm run fed:quorum
+  docker images
+
+  # If enough time have passed and there are still containers not ready then
+  # just assume that they are in a crash loop and abort CI run.
+  iterationCount=1
+  while docker ps | grep "starting\|unhealthy"; do
+    if [ "$iterationCount" -gt 20 ]; then
+      false;
+    fi
+    iterationCount=$[$iterationCount +1]
+    sleep 15; echo; date;
+  done
+
+  docker ps -a
+  sleep 120
```

Comment:
12 looks enough :)

Reply:
> 12 looks enough :)

It did to me as well, until it wasn't when I tested it on weaker hardware. There's probably a slightly lower value for delay that still works most of the time, but this was much quicker for now and given that it is a band-aid regardless of how long the wait is, I figured it's not worth doing a binary search for the "optimal" delay time. Instead what I focused on is having the integration points (mostly the REST API calls and the Web3 transactions) implemented in a more resilient fashion which - if done right - then eliminates the need for any hacky delays completely. This person says explains it much better: https://medium.com/@kelseyhightower/12-fractured-apps-1080c73d481c